subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/DefaultArtifactTransforms.java

```diff
@@ -0,0 +1,219 @@
+        if (attributes == null || attributes.isEmpty()) {
+            return visitor;
+        }
+        return new ArtifactTransformingVisitor(visitor, attributes.asImmutable());
+    }
+
+    private class AttributeMatchingVariantSelector<T extends HasAttributes> implements Transformer<T, Collection<? extends T>> {
+        private final AttributeContainer attributes;
+
+        private AttributeMatchingVariantSelector(AttributeContainer attributes) {
+            this.attributes = attributes;
+        }
+
+        @Override
+        public String toString() {
+            return "Variant selector for " + attributes;
+        }
+
+        @Override
+        public T transform(Collection<? extends T> variants) {
+            // Note: This algorithm is a placeholder only. Should deal with ambiguous matches
+            if (attributes.isEmpty()) {
+                return variants.iterator().next();
+            }
+
+            // Note: This algorithm is a placeholder only. Should deal with ambiguous matches
+            T canTransform = null;
+            for (T variant : variants) {
+                AttributeContainer variantAttributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
+                if (artifactTransformRegistrations.areMatchingAttributes(variantAttributes, this.attributes)) {
+                    return variant;
+                }
+                if (artifactTransformRegistrations.getTransform(variantAttributes, this.attributes) != null) {
+                    canTransform = variant;
+                }
+            }
+            return canTransform;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof AttributeMatchingVariantSelector)) {
+                return false;
+            }
+            AttributeMatchingVariantSelector<?> that = (AttributeMatchingVariantSelector<?>) o;
+            return Objects.equal(attributes, that.attributes);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(attributes);
+        }
+    }
+
+    private class ArtifactTransformingVisitor implements ArtifactVisitor {
+        private final ArtifactVisitor visitor;
+        private final AttributeContainerInternal attributes;
+
+        private ArtifactTransformingVisitor(final ArtifactVisitor visitor, @Nullable final AttributeContainerInternal attributes) {
+            this.visitor = visitor;
+            this.attributes = attributes;
+        }
+
+        @Override
+        public void visitArtifact(final ResolvedArtifact artifact) {
+            List<ResolvedArtifact> transformResults = artifactTransformRegistrations.getTransformedArtifacts(artifact, attributes);
+            if (transformResults == null) {
+                if (artifactTransformRegistrations.areMatchingAttributes(artifact.getAttributes(), this.attributes)) {
```

Comment:
This looks like we actually have a three-step matching right now:

1. Configuration
2. Variant
3. Artifact

Shouldn't we collapse the last two?